Resolve #1118 -- Add/removeBuff Change

cabeca1143 opened the pull request:
-Changed so AddBuff and RemoveBuff works with IAttackableUnits instead of IObjAiBase.

-Updated Akali's QE Scripts & Introduced her Q mark, to support the Add/RemoveBuff changes.

-Fixed Broken Blitzcrank Q2 Buff Script

Resolve #1118

cabeca1143 committed:
Fixed Blitzcrank's Broken buff Script

cabeca1143 committed:
AddBuff & RemoveBuff Update

Changed IObjAiBase to IAttackableUnit in order to be possible to Add and Remove buffs from minions.
-Akali's QE rewritten to support that change

lzardy commented on Content/LeagueSandbox-Scripts/Buffs/Akali/QMark.cs:
```diff
@@ -0,0 +1,44 @@
+
+        public void OnActivate(IAttackableUnit unit, IBuff buff, ISpell ownerSpell)
+        {
+            pbuff = AddParticleTarget(ownerSpell.CastInfo.Owner, unit, "akali_markOftheAssasin_marker_tar_02.troy", unit, buff.Duration);
```
AkaliMota uses `akali_markoftheassasin_marker_tar` as well as `akali_markoftheassasin_marker_tar_02`

danil179 commented on GameServerCore/Domain/GameObjects/IBuff.cs:
```diff
@@ -43,7 +43,7 @@ public interface IBuff : IStackable, IUpdate
         /// Unit which applied this buff to its target.
         /// </summary>
-        IObjAiBase SourceUnit { get; }
+        IAttackableUnit SourceUnit { get; }
```
`SourceUnit` can be IAttackableUnit?

lzardy replied:
I would not recommend this change as it sets the precedent that AttackableUnit can affect other units, which imo should only be true for position/movement. After reverting this change, make sure to remove the ObjAiBase cast in the AkaliMota buff.

cabeca1143 replied:
That had to be changed in order for the changes in "from" in AddBuff and "target" in RemoveBuff work properly.
Any suggestion on another path that could be taken if it's so problematic?

lzardy replied:
You should be able to revert the `from` in `AddBuff` to `ObjAiBase`.

cabeca1143 replied:
Reverted all IObjAiBases besides in the RemoveBuff and fixed the issues in other scripts.

lzardy commented on Content/LeagueSandbox-Scripts/Champions/Akali/Q.cs:
```diff
@@ -26,6 +34,7 @@ public void OnDeactivate(IObjAiBase owner, ISpell spell)
 
         public void OnSpellPreCast(IObjAiBase owner, ISpell spell, IAttackableUnit target, Vector2 start, Vector2 end)
         {
+            ApiEventManager.OnSpellMissileHit.AddListener(this, new KeyValuePair<ISpell, IObjAiBase>(spell, owner), TargetExecute, true);
```
Move this to OnActivate and change `true` to `false`. Otherwise, if you attempt to fire two of the projectiles, a single projectile will do damage worth two.

danil179 reviewed (commented):
I didn't really checked the scripts. They seem fine to me. 

I have a problem with the change of `SourceUnit` type (from `IOnjAiBase` to `IAttackableUnit` in `IBuff`, is it needed? 

I agree that if Buff is in `IAttackableUnit` then all the buff related API apply on it as well.

cabeca1143 committed:
Update

-Added missing Mark particle (didn't notice any visual change tho)

-Fixed misplaced Listener on Akali's Q

-Reverted Back all the changes from IObjAiBase to IAttackableUnit, besides in the RemoveBuff function. (I could swear there was a reason I did those changes... Well, apparently not)

danil179 reviewed (approved):
Approved.
